### DISCUSSION TITLE
Masking and formatting using annotations

### DISCUSSION THREAD

**Discussion by pecigonzalo** [logfellow/logstash-logback-encoder#824] [Ideas]
Hi! Im looking to implement masking of some fields for Arguments and MDC given the origin object annotations. Something along the lines of what is decribed in https://gkovan.medium.com/pci-and-pii-compliance-when-logging-data-in-digital-transformation-projects-7739bab159a6

I've gave it a go using custom Providers and the likes, but it seems that from the provider perspective I can only access maps/strings, but not the original objects, so I can do custom actions based on the annotations.

This would be my sample data class and annotation
```kotlin
@Target(
    AnnotationTarget.FIELD,
)
@Retention(AnnotationRetention.RUNTIME)
@MustBeDocumented
annotation class Mask(val mask: String = "*****")

@Serializable
data class Customer(
    val id: String,
    val firstName: String,
    val lastName: String,
    @Mask() val email: String
)
```

Any ideas or pointers? Anyone has implemented something similar?

**Comment by brenuart** [comment]
Hi @pecigonzalo ,

That would be a nice feature indeed. A few questions tho...
The behaviour of the @Mask annotation is quite obvious when put on a String field. But what would you expect in the following scenarios:

- Annotation on a Collection
```
@Mask Collection<String> collectionOfString;
```
Would you expect the entire collection field to be masked - similar to the current masking behaviour when the path points to the field holding the collection? Example:
```
"collectionOfString": "****"
```

Or maybe the @Mask annotation should accept a path pattern that would be evaluated starting at the annotated field? The following example would then mask only the second entry of the collection (index is 0 based):
```
@Mask(path="1") Collection<String> collectionOfString;
```
This would produce a JSON similar to:
```
"collectionOfString" : [
   "first string",
   "****",
   "third string"
]
```

- Annotation on a Pojo

Same question when the annotation is put on a field holding a Pojo..
```
@Mask Pojo mojo;
```

**Comment by pecigonzalo** [comment]
> Would you expect the entire collection field to be masked

I think this would be a good option yeah. It could be interesting to mask on each item (eg a list of email), but this might prove a bit complex both implementation wise and processing wise. If we assume a collection is of a type of something, we most likely want all of the collection masked in the same fashion, so `@Mask(path="1")` would not be my choice. I would rather mask all the items in the same way, or just place a string as in your example.

> Annotation on a Pojo

I would not traverse the Pojo, if the Pojo needs masking the it should be defined in the data class of that pojo.

PS: Sorry for the long wait here!
